python/ray/llm/_internal/serve/deployments/routers/prefix_tree.py

```diff
@@ -0,0 +1,379 @@
+from ray import serve
+import time
+from threading import RLock
+from typing import Optional, List, Tuple, Dict, Set
+
+
+class Node:
+    """
+    Node in a prefix tree that tracks tenant access time.
+
+    Each node represents a segment of text and can belong to multiple tenants.
+    """
+
+    def __init__(self, text: str = "", parent: Optional["Node"] = None):
+        self.text: str = text
+        self.parent: Optional["Node"] = parent
+        self.children: Dict[str, "Node"] = {}  # Maps char -> Node
+        self.tenant_last_access_time: Dict[
+            str, int
+        ] = {}  # Maps tenant -> timestamp in ms (int)
+
+    def to_string(self) -> str:
+        return f"Node(text='{self.text}', parent={self.parent}, children={self.children}, tenant_last_access_time={self.tenant_last_access_time})"
+
+
+@serve.deployment(name="TreeDeployment")
+class PrefixTree:
+    """
+    Thread-safe multi-tenant prefix tree (approximate radix tree).
+
+    Features:
+    1. Stores data for multiple tenants in the same tree structure
+    2. Node-level locking for concurrent access
+    3. Leaf LRU eviction based on tenant access time
+    """
+
+    def __init__(self) -> None:
+        self.lock: RLock = RLock()
+        self.root: Node = Node()
+        self.tenants: Set[str] = set()
+        self.tenant_char_count: Dict[str, int] = {}
+        self.tenant_nodes: Dict[str, Set[Node]] = {}
+
+    def reset(self) -> None:
+        """Reset the tree to an empty state."""
+        with self.lock:
+            self.root = Node()
+            self.tenants = set()
+            self.tenant_char_count = {}
+            self.tenant_nodes = {}
+
+    def to_dict(self) -> Dict:
+        return {
+            "root": self.root,
+            "tenants": self.tenants,
+            "tenant_char_count": self.tenant_char_count,
+            "tenant_nodes": self.tenant_nodes,
+        }
+
+    def to_string(self) -> str:
+        return f"PrefixTree(root={self.root.__str__()}, tenants={self.tenants}, tenant_char_count={self.tenant_char_count}, tenant_nodes={self.tenant_nodes})"
+
+    @staticmethod
```

Comment:
asked AI to optimize this code. The suggestion was: 

```
@staticmethod
def shared_prefix_count(a: str, b: str) -> int:
    """Count the number of shared characters at the beginning of two strings."""
    return len(os.path.commonprefix([a, b]))
```

This uses C under the hood and does not have the python overhead. So in super long strings of A and B should be faster.

Reply:
also make this a private function.